### PR TITLE
fix: scroll to right position on text insertion

### LINKED ISSUE #7351: [HOLD for payment 2022-02-09] Large paragraph goes beyond the bottom of the visible input field and doesn't scroll
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app 
2. Navigate to any chat 
3. Paste a large text 

## Expected Result:
When I paste a large paragraph into the input field in NewDot, where that paragraph goes beyond the bottom of the visible input field, I would expect it to scroll down to the end of my input. 

## Actual Result:
When I paste a large paragraph, It does not scroll down, leaving user to wonder (until scroll down) whether the whole thing pasted or just the first line that is visible.

## Workaround:
Unknown 

## Platform:

Where is this issue occurring?

- Desktop App
 
**Version Number:** 1.1.30-3
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/150466401-362d3003-baef-4a92-9524-ed7387741a7a.mov

**Expensify/Expensify Issue URL:**
**Issue reported by:** @rafecolton 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1642539528151600

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

### LINKED ISSUE #8663: [$250] mWeb/Chrome - Text Insertion point goes out of the sight on typing in the middle of the paragraph - reported by @Tushu17
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
 1. Go Chat
 2. Write 10-12 line text
 3. Type some words on the 4th-5th line.

## Expected Result:
Text insertion point should remain at a constant position.

## Actual Result:
It kept moving in and out of sight.

## Workaround:
unknown. 

## Platform:

Where is this issue occurring?

- Mobile Web (Chrome)

**Version Number:** 1.1.49-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/163698343-dcc08aef-380e-40a7-ab58-27bbd873cd6f.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @Tushu17
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1646931771824919

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

### PR DESCRIPTION
**Pull request by b1tjoy**
### Details
Text Insertion point goes out of the sight on typing in the middle of the paragraph. This PR fix the issue. And in order to fix #8663, we have to delete one line of code, which was add to fix #7351 previous, so this PR alse fix that one.

**Why move textInput.blur() and textInput.focus() from ReportActionCompose.js to Composer/index.js?**

Because these two lines was add to fix #7351, but the issue only affects Web, mWeb and Desktop platform, and only in pasting text copied from HTML page scenario, if we keep the code in ReportActionCompose.js, it will cause other issues, such as:

1. Virtual keyboard keeps collapse and pop out on Android native when we type text in input field
2. Input field keeps flash on Web when we type text

So I moved these codes to Composer/index.js:handlePasteHTML().

### Fixed Issues
$ [#8663](https://github.com/Expensify/App/issues/8663)

### Tests
**Test case for #8663:**

1. Open Expensify account
2. Select any chat room in your list
3. Write multiple lines of text, 20+ lines for web, 10+ lines for mWeb
4. Type some words on the 4th-5th line
5. Verify text insertion point does not move out of sight

**Test case for #7351:**

1. Open Expensify account
2. Select any chat room in your list
3. Paste a large paragraph copy from HTML page, 20+ lines for web, 10+ lines for mWeb
4. Verify the scroll bar scroll down to the end of input field and text insertion point remains in sight

**Test case for paste on selection, a more complex scenario:**

1. Open Expensify account
2. Select any chat room in your list
3. Write multiple lines of text, 20+ lines for web, 10+ lines for mWeb
4. Copy a large parapragh from HTML page
5. Select some text from chat input field and paste
6. Verify the text insertion point remains in sight

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
**Test case for #8663:**

1. Open Expensify account
2. Select any chat room in your list
3. Write multiple lines of text, 20+ lines for web, 10+ lines for mWeb
4. Type some words on the 4th-5th line
5. Verify text insertion point does not move out of sight

**Test case for #7351:**

1. Open Expensify account
2. Select any chat room in your list
3. Paste a large paragraph copy from HTML page, 20+ lines for web, 10+ lines for mWeb
4. Verify the scroll bar scroll down to the end of input field and text insertion point remains in sight

**Test case for paste on selection, a more complex scenario:**

1. Open Expensify account
2. Select any chat room in your list
3. Write multiple lines of text, 20+ lines for web, 10+ lines for mWeb
4. Copy a large parapragh from HTML page
5. Select some text from chat input field and paste
6. Verify the text insertion point remains in sight

- [ ] Verify that no errors appear in the JS console

#### Web

Test case for #8663

https://user-images.githubusercontent.com/103875612/169700492-d9e52c7b-ab5d-4847-b847-299fa5f6ec85.mp4

Test case for #7351 

https://user-images.githubusercontent.com/103875612/169700531-d14320dd-7c0c-4607-aaf3-1dfcf1dd62fb.mp4

Test case for paste on selection, a more complex scenario

https://user-images.githubusercontent.com/103875612/169700559-613c0134-c800-483b-85f3-d4d9a9a787df.mp4

#### Mobile Web

Test case for #8663 Android/Chrome

https://user-images.githubusercontent.com/103875612/169700750-5f1b47b4-1457-4614-ad53-29078030ac9f.mp4

Test case for #7351 Android/Chrome

https://user-images.githubusercontent.com/103875612/169700808-b8ae06e4-c57a-4c4e-bf6c-d3d6740b3cab.mp4

Test case for paste on selection, a more complex scenario Android/Chrome

https://user-images.githubusercontent.com/103875612/169700962-36406b3b-8b2a-4395-ab58-09dc92995398.mp4

Test case for #8663 iOS/Safari

https://user-images.githubusercontent.com/103875612/169700971-a4263165-ee24-4e23-a8e2-a0dc7aaf2880.mp4

Test case for #7351 iOS/Safari

https://user-images.githubusercontent.com/103875612/169700981-83b64be4-0594-4465-a6ca-7bccdf120e1e.mp4

Test case for paste on selection, a more complex scenario iOS/Safari

https://user-images.githubusercontent.com/103875612/169700989-2d811697-1069-42a2-b723-327c46136a77.mp4

#### Desktop

Test case for #8663

https://user-images.githubusercontent.com/103875612/169700994-36b4cff3-9db3-4d48-a02c-1d8894b27203.mp4

Test case for #7351 

https://user-images.githubusercontent.com/103875612/169701000-c6b36e79-7aef-4e25-8285-2327e07ef240.mp4

Test case for paste on selection, a more complex scenario

https://user-images.githubusercontent.com/103875612/169701006-24da5b67-dfee-4a0a-a102-be7d6d9cfeb2.mp4

#### iOS

Test case for #8663

https://user-images.githubusercontent.com/103875612/169701014-1ca7fe05-46ae-4156-a39c-0c145ad67490.mp4

Test case for #7351 

https://user-images.githubusercontent.com/103875612/169701023-bd8bf4b4-61fb-4efe-842c-73e69ad5c314.mp4

Test case for paste on selection, a more complex scenario

https://user-images.githubusercontent.com/103875612/169701028-01dfe849-2579-4b42-a18b-04319c01c041.mp4

#### Android

Test case for #8663

https://user-images.githubusercontent.com/103875612/169701039-83a3bf1f-affe-442a-ae4e-b6a95e7c6589.mp4

Test case for #7351 

https://user-images.githubusercontent.com/103875612/169701061-fa334d2f-0f1d-4eef-b906-2a074ff8238d.mp4

Test case for paste on selection, a more complex scenario

https://user-images.githubusercontent.com/103875612/169701072-28f34a89-e205-442c-925f-438c09c1ac57.mp4

### CODE REVIEW

**Comment by b1tjoy** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/components/Composer/index.js [hunk 1]
```diff
@@ -244,6 +244,8 @@ class Composer extends React.Component {
         try {
             document.execCommand('insertText', false, markdownText);
             this.updateNumberOfLines();
+            this.textInput.blur();
```
Comment (parasharrajat) [inline src/components/Composer/index.js:247]: A comment will be useful.

**Review by parasharrajat — CHANGES_REQUESTED** [review]
Thanks for the detailed PR.

**Comment by parasharrajat** [comment]
The review of this PR will be delayed. I need to run a few tests. The current solution is also a hack to fix the cursor. I will try to find a better solution to the problem.

**Comment by b1tjoy** [comment]
> The review of this PR will be delayed. I need to run a few tests. The current solution is also a hack to fix the cursor. I will try to find a better solution to the problem.

I agree with you, if we use a hack to fix one issue at the current moment, it may lead to another issue in the future. Always find the root cause.

First I thought it was related to react-native-web, but after some digging, it may be problem with Chrome. I write snippets with basic JS to show how document.execCommand('insertText') works:

[Text Insertion Example](https://jsbin.com/kazoder/edit?html,js,output)

If you run with Chrome, when you paste a large paragraph, the text insertion point goes out of sight. But if you run the example with Firefox, everything works fine.

**Comment by parasharrajat** [comment]
No update yet. But I will share some update very soon.

**Comment by parasharrajat** [comment]
@b1tjoy Thank you for your patience. I was looking into your other proposals. And it seems that this https://github.com/Expensify/App/issues/8663#issuecomment-1116002461 could be the next potential solution. 

You mentioned that this is working fine. Was there any issues with it?

**Comment by b1tjoy** [comment]
@parasharrajat When we select text from bottom to top, then paste the content copy from html page, the text insertion point moves out of sight.

https://user-images.githubusercontent.com/103875612/173195610-ca47bf34-9d97-418b-9b4b-06ffd336facd.mp4

**Comment by parasharrajat** [comment]
Ok, I am trying to determine the expected behavior when we paste text in textArea. There are a few case possible for pasting. I will try to list them down and test your solution over it.

**Comment by b1tjoy** [comment]
> Ok, I am trying to determine the expected behavior when we paste text in textArea. There are a few case possible for pasting. I will try to list them down and test your solution over it.

@parasharrajat Possible cases for pasting I thought of:

1. textArea is empty, paste text more than 20 lines
2. textArea not empty, no selection in textArea, paste in the beginning, middle or end of textArea
3. textArea not empty, select partial text from top to bottom in the beginning, middle or end of textArea, then paste
4. textArea not empty, select partial text from bottom to top in the beginning, middle or end of textArea, then paste

I don't want to be bothersome and I know you are pretty busy, could you please take some time to review this PR for me? As a new contributor, I am not allowed to work on other issues for now, although I have submitted [good proposal](https://github.com/Expensify/App/issues/9183#issuecomment-1159452644). Thanks a lot! :)

**Comment by parasharrajat** [comment]
Sorry for the delay. I just kind of lost this one. Adding it to my list now. I will get to the bottom of it in a day or two. 

Anyways, you should be good for new tasks. You can ping again on the same issue.

**Comment by parasharrajat** [comment]
Ok, I tested your https://github.com/Expensify/App/issues/8663#issuecomment-1116002461 proposal and it seems to be working. Could you do the following and run a few tests? Otherwise, I can go with the existing solution. 

1. Add a state update callback as second param and add these statements to the callback. 
```js
 const heightToBottom = this.textInput.scrollHeight - paddingTopAndBottom - this.textInput.scrollTop;
                this.textInput.scrollTop = this.textInput.scrollHeight - paddingTopAndBottom - heightToBottom;
```

Now run your tests.

**Comment by parasharrajat** [comment]
Also, please merge main.

**Comment by b1tjoy** [comment]
I don't think it's going to work. The main idea of https://github.com/Expensify/App/issues/8663#issuecomment-1116002461 is to store the height from the insertion point to the last line of textArea before document.execCommand('insertText'), after text insertion we can calculate scrollTop by the stored value and manually scroll the textArea to the appropriate position. This works fine for the following cases:

1. textArea is empty, paste text more than 20 lines
2. textArea not empty, no selection in textArea, paste in the beginning, middle or end of textArea
3. textArea not empty, select partial text from top to bottom in the beginning, middle or end of textArea, then paste

If we have 40 lines text for example, line height is 20, `maxLines = 16`, select from line 11 to 30, and paste 30 lines text:
```JSX
Before insert:
scrollHeight = 800
selection lines (select from top to bottom) = [11, 30]
insertion point line = 30
visible area lines = [15, 30]
upper lines not visible = [1, 14]
lower lines not visible = [31, 40]
scrollTop = 14 * 20 = 280
heightToBottom = 800 - 280 = 520

After insert:
scrollHeight = 1000
scrollTop = 1000 - 520 = 480 = 24 * 20
upper lines not visible = [1, 24]
visible area lines = [25, 40]
lower lines not visible = [41, 50]
insertion point line = 40 (because we replace line 11-30 with 30 lines text)
```
Insertion point is in the visible area and everything works fine.

But in some edge case it doesn't work:

- textArea not empty, select partial text from bottom to top (**more than 16 lines and make scroll bar goes backward**) in textArea, then paste

If we have 40 lines text for example, line height is 20, `maxLines = 16`, select from line 30 to 11, and paste 30 lines text:
```JSX
Before insert:
scrollHeight = 800
selection lines (select from bottom to top) = [30, 11]
insertion point line = 30
visible area lines = [11, 26]
upper lines not visible = [1, 10]
lower lines not visible = [27, 40]
scrollTop = 10 * 20 = 200
heightToBottom = 800 - 200 = 600

After insert:
scrollHeight = 1000
scrollTop = 1000 - 600 = 400 = 20 * 20
upper lines not visible = [1, 20]
visible area lines = [21, 36]
lower lines not visible = [37, 50]
insertion point line = 40 (because we replace line 11-30 with 30 lines text)
```
Insertion point is not in the visible area.

I think we can combine these two proposals, use the calculation one to handle regular case, and use blur() & focus() one to handle select backward case. We can add a variable to track if user scroll backward, if user scroll backward and selection is not empty, call blur() & focus() to keep insertion point insight, otherwise calculate scrollTop and manually scroll to appropriate position. How about that?

**Comment by parasharrajat** [comment]
A slight difference from your proposal is that I am not tracking the scroll before paste. Instead, I am relying on the reset of `numberOfLines` to `1`. 

Anyways, I see your analysis. Testing the current solution in this PR. I would say that if are not able to set a manual scroll position then this solution is fine. I will open the discussion on slack to get more eyes before moving ahead with the PR. 

It would be done by tomorrow.

**Comment by parasharrajat** [comment]
Meanwhile please merge main into this PR.

**Comment by parasharrajat** [comment]
Yeah, I saw that the current solution of the PR has better results. I will open the discussion today.

File: src/components/Composer/index.js [hunk 2]
```diff
@@ -239,6 +239,10 @@ class Composer extends React.Component {
         try {
             document.execCommand('insertText', false, markdownText);
             this.updateNumberOfLines();
+            // Manually make text input scroll to appropriate position, otherwise the text insertion
+            // point will goes out of sight when we paste a large paragraph on Chrome or Desktop
```
Comment (parasharrajat) [inline src/components/Composer/index.js:243]:
```suggestion
            // Pointer will go out of sight when a large paragraph is pasted on the web. Refocusing the input keeps the cursor in view.
```

File: src/components/Composer/index.js [hunk 3]
```diff
@@ -239,6 +239,10 @@ class Composer extends React.Component {
         try {
             document.execCommand('insertText', false, markdownText);
             this.updateNumberOfLines();
+            // Manually make text input scroll to appropriate position, otherwise the text insertion
```
Comment (parasharrajat) [inline src/components/Composer/index.js:242]:
```suggestion

```

**Review by parasharrajat — CHANGES_REQUESTED** [review]
@b1tjoy Please merge main into this.

**Review by parasharrajat — APPROVED** [review]
LGTM. 

cc: @pecanoro 

#### PR Reviewer Checklist
- [x] I verified the PR has a small number of commits behind `main`
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the testing environment is mentioned in the test steps
- [x] I verified testing steps cover success & fail scenarios (if applicable)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors related to changes in this PR
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified comments were added when the code was not self explanatory
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files (if applicable)
    - [x] I verified proper naming convention for platform-specific files was followed (if applicable)
    - [x] I verified [style guidelines](https://github.com/Expensify/App/blob/main/STYLING.md) were followed
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components are not impacted by changes in this PR (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified the UI performance was not affected (the performance is the same than `main` branch)
- [x] If a new component is created I verified that a similar component doesn't exist in the codebase

**Review by pecanoro — APPROVED** [review]

**Pull request merged**